pkg/virt-controller/watch/snapshot/restore.go

```diff
@@ -0,0 +1,537 @@
+		updateRestoreCondition(vmRestoreOut, newReadyCondition(corev1.ConditionFalse, "Waiting for new PVCs"))
+	}
+
+	return 0, ctrl.doUpdate(vmRestoreIn, vmRestoreOut)
+}
+
+func (ctrl *VMRestoreController) doUpdateError(original, updated *snapshotv1.VirtualMachineRestore, err error) error {
+	ctrl.Recorder.Eventf(
+		updated,
+		corev1.EventTypeWarning,
+		restoreErrorEvent,
+		"VirtualMachineRestore encountered error %s",
+		err.Error(),
+	)
+
+	updateRestoreCondition(updated, newProgressingCondition(corev1.ConditionFalse, err.Error()))
+	updateRestoreCondition(updated, newReadyCondition(corev1.ConditionFalse, err.Error()))
+	if err2 := ctrl.doUpdate(original, updated); err2 != nil {
+		return err2
+	}
+
+	return err
+}
+
+func (ctrl *VMRestoreController) doUpdate(original, updated *snapshotv1.VirtualMachineRestore) error {
+	if !reflect.DeepEqual(original, updated) {
+		if _, err := ctrl.Client.VirtualMachineRestore(updated.Namespace).Update(updated); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *VMRestoreController) reconcileVolumeRestores(vmRestore *snapshotv1.VirtualMachineRestore) (bool, error) {
+	content, err := ctrl.getSnapshotContent(vmRestore)
+	if err != nil {
+		return false, err
+	}
+
+	var restores []snapshotv1.VolumeRestore
+	for _, vb := range content.Spec.VolumeBackups {
+		found := false
+		for _, vr := range vmRestore.Status.Restores {
+			if vb.VolumeName == vr.VolumeName {
+				restores = append(restores, vr)
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			if vb.VolumeSnapshotName == nil {
+				return false, fmt.Errorf("VolumeSnapshotName missing %+v", vb)
+			}
+
+			vr := snapshotv1.VolumeRestore{
+				VolumeName:                vb.VolumeName,
+				PersistentVolumeClaimName: restorePVCName(vmRestore, vb.VolumeName),
+				VolumeSnapshotName:        *vb.VolumeSnapshotName,
+			}
+			restores = append(restores, vr)
+		}
+	}
+
+	if !reflect.DeepEqual(vmRestore.Status.Restores, restores) {
+		if len(vmRestore.Status.Restores) > 0 {
+			log.Log.Object(vmRestore).Warning("VMRestore in strange state")
+		}
+
+		vmRestore.Status.Restores = restores
+		return true, nil
+	}
+
+	createdPVC := false
+	for i, restore := range restores {
+		pvc, err := ctrl.getPVC(vmRestore.Namespace, restore.PersistentVolumeClaimName)
+		if err != nil {
+			return false, err
+		}
+
+		if pvc == nil {
+			backup := content.Spec.VolumeBackups[i]
+			if err = ctrl.createRestorePVC(vmRestore, backup, restore); err != nil {
+				return false, err
+			}
+			createdPVC = true
+		}
+	}
+
+	return createdPVC, nil
+}
+
+func (t *vmRestoreTarget) Ready() (bool, error) {
+	log.Log.Object(t.vmRestore).V(3).Info("Checking VM ready")
+
+	vmiKey, err := controller.KeyFunc(t.vm)
+	if err != nil {
+		return false, err
+	}
+
+	_, exists, err := t.controller.VMIInformer.GetStore().GetByKey(vmiKey)
+	if err != nil {
+		return false, err
+	}
+
+	return !exists, nil
```

Comment:
would it make sense to check the runStrategy and running options on the VM spec as well to ensure the VM is meant to be shutdown as well?

Reply:
won't hurt